Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean

```diff
@@ -95,27 +95,143 @@ abbrev FunctionField : Type u :=
 noncomputable abbrev mk : R[X][Y] →+* W.CoordinateRing :=
   AdjoinRoot.mk W.polynomial
 
+-- Porting note: added `classical` explicitly
+/-- The basis $\{1, Y\}$ for the coordinate ring $R[W]$ over the polynomial ring $R[X]$. -/
+protected noncomputable def basis : Basis (Fin 2) R[X] W.CoordinateRing := by
+  classical exact (subsingleton_or_nontrivial R).by_cases (fun _ => default) fun _ =>
+    (AdjoinRoot.powerBasis' W.monic_polynomial).basis.reindex <| finCongr W.natDegree_polynomial
+#align weierstrass_curve.coordinate_ring.basis WeierstrassCurve.Affine.CoordinateRing.basis
+
+lemma basis_apply (n : Fin 2) :
+    CoordinateRing.basis W n = (AdjoinRoot.powerBasis' W.monic_polynomial).gen ^ (n : ℕ) := by
+  classical
+  nontriviality R
+  rw [CoordinateRing.basis, Or.by_cases, dif_neg <| not_subsingleton R, Basis.reindex_apply,
+    PowerBasis.basis_eq_pow]
+  rfl
+#align weierstrass_curve.coordinate_ring.basis_apply WeierstrassCurve.Affine.CoordinateRing.basis_apply
+
+-- Porting note: added `@[simp]` in lieu of `coe_basis`
+@[simp]
+lemma basis_zero : CoordinateRing.basis W 0 = 1 := by
+  simpa only [basis_apply] using pow_zero _
+#align weierstrass_curve.coordinate_ring.basis_zero WeierstrassCurve.Affine.CoordinateRing.basis_zero
+
+-- Porting note: added `@[simp]` in lieu of `coe_basis`
+@[simp]
+lemma basis_one : CoordinateRing.basis W 1 = mk W Y := by
+  simpa only [basis_apply] using pow_one _
+#align weierstrass_curve.coordinate_ring.basis_one WeierstrassCurve.Affine.CoordinateRing.basis_one
+
+-- Porting note: removed `@[simp]` in lieu of `basis_zero` and `basis_one`
+lemma coe_basis : (CoordinateRing.basis W : Fin 2 → W.CoordinateRing) = ![1, mk W Y] := by
+  ext n
+  fin_cases n
+  exacts [basis_zero W, basis_one W]
+#align weierstrass_curve.coordinate_ring.coe_basis WeierstrassCurve.Affine.CoordinateRing.coe_basis
+
+variable {W} in
+lemma smul (x : R[X]) (y : W.CoordinateRing) : x • y = mk W (C x) * y :=
+  (algebraMap_smul W.CoordinateRing x y).symm
+#align weierstrass_curve.coordinate_ring.smul WeierstrassCurve.Affine.CoordinateRing.smul
+
+variable {W} in
+lemma smul_basis_eq_zero {p q : R[X]} (hpq : p • (1 : W.CoordinateRing) + q • mk W Y = 0) :
+    p = 0 ∧ q = 0 := by
+  have h := Fintype.linearIndependent_iff.mp (CoordinateRing.basis W).linearIndependent ![p, q]
+  erw [Fin.sum_univ_succ, basis_zero, Fin.sum_univ_one, basis_one] at h
+  exact ⟨h hpq 0, h hpq 1⟩
+#align weierstrass_curve.coordinate_ring.smul_basis_eq_zero WeierstrassCurve.Affine.CoordinateRing.smul_basis_eq_zero
+
+variable {W} in
+lemma exists_smul_basis_eq (x : W.CoordinateRing) :
+    ∃ p q : R[X], p • (1 : W.CoordinateRing) + q • mk W Y = x := by
+  have h := (CoordinateRing.basis W).sum_equivFun x
+  erw [Fin.sum_univ_succ, Fin.sum_univ_one, basis_zero, basis_one] at h
+  exact ⟨_, _, h⟩
+#align weierstrass_curve.coordinate_ring.exists_smul_basis_eq WeierstrassCurve.Affine.CoordinateRing.exists_smul_basis_eq
```

Comment:
Why is this better than
```suggestion
variable {W}

lemma smul (x : R[X]) (y : W.CoordinateRing) : x • y = mk W (C x) * y :=
  (algebraMap_smul W.CoordinateRing x y).symm
#align weierstrass_curve.coordinate_ring.smul WeierstrassCurve.Affine.CoordinateRing.smul

lemma smul_basis_eq_zero {p q : R[X]} (hpq : p • (1 : W.CoordinateRing) + q • mk W Y = 0) :
    p = 0 ∧ q = 0 := by
  have h := Fintype.linearIndependent_iff.mp (CoordinateRing.basis W).linearIndependent ![p, q]
  erw [Fin.sum_univ_succ, basis_zero, Fin.sum_univ_one, basis_one] at h
  exact ⟨h hpq 0, h hpq 1⟩
#align weierstrass_curve.coordinate_ring.smul_basis_eq_zero WeierstrassCurve.Affine.CoordinateRing.smul_basis_eq_zero

lemma exists_smul_basis_eq (x : W.CoordinateRing) :
    ∃ p q : R[X], p • (1 : W.CoordinateRing) + q • mk W Y = x := by
  have h := (CoordinateRing.basis W).sum_equivFun x
  erw [Fin.sum_univ_succ, Fin.sum_univ_one, basis_zero, basis_one] at h
  exact ⟨_, _, h⟩
#align weierstrass_curve.coordinate_ring.exists_smul_basis_eq WeierstrassCurve.Affine.CoordinateRing.exists_smul_basis_eq

variable (W)
```
?

Reply:
I would much prefer finer control on the explicit/implicitness of the variables. When I move things around they sometimes get accidentally introduced/hidden without me noticing. Sometimes this doesn't affect the lemmas in the same file that use that specific lemma (e.g. they're just `rw`s without specifying the arguments), but affects downstream files (e.g. they're `exact`s supplying all the arguments). When I inevitably split this file in the future to match the Jacobian split I don't want to get confused myself.

Reply:
I see, this is a good cause!